### PR TITLE
ext3grep: init at 0.10.2

### PR DESCRIPTION
**Pull request by D3vil0p3r**
## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by AndersonTorres** [comment]
I created a backup from Google Code to GitHub. Wanna try?

https://github.com/museoa/ext3grep

**Comment by AndersonTorres** [comment]
Yep. It is a svn->git conversion.

**Comment by D3vil0p3r** [comment]
I would keep the original source currently. @h7x4 what do you think?

File: pkgs/by-name/ex/ext3grep/package.nix [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  patches = [
+    ./ext3grep-build.patch
+    ./ext3grep-e2fsprogs-1.44.1.patch
+  ];
+
+  buildPhase = ''
+    ./configure --prefix=$out
+    make
+  '';
+
```
[the remark this answers is not in the document]
Comment (D3vil0p3r) [reply pkgs/by-name/ex/ext3grep/package.nix:27]: @a-n-n-a-l-e-e if I remove it and I use the autoreconfHook, aclocal generates an error that is described here: https://www.mail-archive.com/debian-bugs-dist@lists.debian.org/msg1629458.html

Did you test it by removing that piece of code?

[the remark this answers is not in the document]
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ex/ext3grep/package.nix:27]: yes, i tested it. don't add autoreconfHook.

File: pkgs/by-name/ex/ext3grep/package.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    url = "https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/ext3grep/ext3grep-${finalAttrs.version}.tar.gz";
+    hash = "sha256-WG8+k50v/XgvbwBrgaPfLcR3xtoD8h7biGDFxPcZjz4=";
+  };
+
+  nativeBuildInputs = [ e2fsprogs ];
+
+  patches = [
+    ./ext3grep-build.patch
+    ./ext3grep-e2fsprogs-1.44.1.patch
+  ];
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ex/ext3grep/package.nix:21]: these patches look like they might be from the debian distribution?
use fetchpatch to pull them, should be able to access them from
https://salsa.debian.org/pkg-security-team/ext3grep/-/tree/debian/master/debian/patches?ref_type=heads
and can grep nixpkgs for `salsa.debian.org` for some examples.

File: pkgs/by-name/ex/ext3grep/package.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    ./configure --prefix=$out
+    make
+  '';
+
+  meta = with lib; {
+    description = "A tool to investigate an ext3 file system for deleted content and possibly recover it";
+    mainProgram = "ext3grep";
+    homepage = "https://code.google.com/archive/p/ext3grep/";
+    maintainers = with maintainers; [ d3vil0p3r ];
+    platforms = platforms.unix;
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ex/ext3grep/package.nix:33]:
```suggestion
    platforms = platforms.unix;
```
the darwin builds are failing -- should this be narrowed to `platforms.linux` ? perhaps the debian patches for clang would fix the build.

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 2 above)
Comment (D3vil0p3r) [reply pkgs/by-name/ex/ext3grep/package.nix:21]: I retrieved those two files from [Arch Linux repository](https://gitlab.archlinux.org/archlinux/packaging/packages/ext3grep) but I cannot find the same patches in that Debian link. Should do apply also the salsa.debian patches?

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 3 above)
Comment (D3vil0p3r) [reply pkgs/by-name/ex/ext3grep/package.nix:33]: Currently I don't have a darwin machine. Can you test it by applying that clang patch file, so, if it works, I can add it to the list of file patches?

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 2 above)
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ex/ext3grep/package.nix:21]: they are the first two patches listed in the debian [link](https://salsa.debian.org/pkg-security-team/ext3grep/-/tree/debian/master/debian/patches?ref_type=heads)
1. https://salsa.debian.org/pkg-security-team/ext3grep/-/blob/debian/master/debian/patches/001_fix-ftbfs-e2fsprogs_1.42-WIP-702.diff?ref_type=heads
2. https://salsa.debian.org/pkg-security-team/ext3grep/-/blob/debian/master/debian/patches/002_remove_i_dir_acl.diff?ref_type=heads

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 3 above)
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ex/ext3grep/package.nix:33]: you can apply the patches and see if the CI succeeds or fails.

Comment (D3vil0p3r) [reply pkgs/by-name/ex/ext3grep/package.nix:33]: Oh true. I am by smartphone and I didn't see the ofborg `
ext3grep, ext3grep.passthru.tests on <arch>-darwin` failing. Once I back home, I apply the required changes ^^.

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 2 above)
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ex/ext3grep/package.nix:21]: > I retrieved those two files from [Arch Linux repository](https://gitlab.archlinux.org/archlinux/packaging/packages/ext3grep) but I cannot find the same patches in that Debian link. Should do apply also the salsa.debian patches?

you can also use fetchpatch to pull them from the arch repo too. there are likely examples in nixpkgs if you grep for the base url.

Comment (AndersonTorres) [reply pkgs/by-name/ex/ext3grep/package.nix:21]: If those patches are "general purpose", then they should be "upstreamed".

File: pkgs/by-name/ex/ext3grep/package.nix [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  src = fetchurl {
+    url = "https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/ext3grep/ext3grep-${finalAttrs.version}.tar.gz";
+    hash = "sha256-WG8+k50v/XgvbwBrgaPfLcR3xtoD8h7biGDFxPcZjz4=";
+  };
+
+  nativeBuildInputs = [ e2fsprogs ];
+
+  patches = [
+    (fetchpatch {
+      url = "https://salsa.debian.org/pkg-security-team/ext3grep/-/raw/debian/master/debian/patches/001_fix-ftbfs-e2fsprogs_1.42-WIP-702.diff";
```
Comment (a-n-n-a-l-e-e) [inline pkgs/by-name/ex/ext3grep/package.nix:21]:
```suggestion
      url = "https://salsa.debian.org/pkg-security-team/ext3grep/-/raw/6150141b945dd12240629c07ada0a033b275baeb/debian/patches/001_fix-ftbfs-e2fsprogs_1.42-WIP-702.diff";
```

**Review by ghost — COMMENTED** [review]
need to use the permalink for the patches in the case they get modified. i fixed the first  patch, the others just need to replace `debian/master` with the commit sha. commit sha can be found from the permalink or when looking at the patch can be copied via a button on the upper right.

~if the build still fails on darwin can wither mark it as broken in the`meta` section if you think it could work but just needs some fixes somewhere. or just use `platforms.linux` if you think just linux is the target platform.~ (see previous comment)

File: pkgs/by-name/ex/ext3grep/package.nix (hunk 3 above)
Comment (a-n-n-a-l-e-e) [reply pkgs/by-name/ex/ext3grep/package.nix:47]: ok -- it looks like the [failures](https://logs.ofborg.org/?key=nixos/nixpkgs.288134&attempt_id=e407ad98-9edc-40cb-8598-d9334114e489) on darwin

```console
In file included from custom.cc:28:
./ext3.h:107:40: error: use of undeclared identifier 'i_uid_low'; did you mean 'uid_low'?
    __u16 uid_low(void) const { return i_uid_low; }
                                       ^~~~~~~~~
                                       uid_low
./ext3.h:107:11: note: 'uid_low' declared here
    __u16 uid_low(void) const { return i_uid_low; }
          ^
./ext3.h:107:40: error: reference to non-static member function must be called; did you mean to call it with no arguments?
    __u16 uid_low(void) const { return i_uid_low; }
                                       ^~~~~~~~~
[...]
```

 are due to defines in `ext2_fs.h` in the `e2fsprogs` package and they are ifdefed for linux or kernel only
```c
#if defined(__KERNEL__) || defined(__linux__)
#define i_reserved1     osd1.linux1.l_i_reserved1
#define i_frag          osd2.linux2.l_i_frag
#define i_fsize         osd2.linux2.l_i_fsize
#define i_uid_low       i_uid
#define i_gid_low       i_gid
#define i_uid_high      osd2.linux2.l_i_uid_high
#define i_gid_high      osd2.linux2.l_i_gid_high
```

so this should just be `platforms.linux`

**Review by ghost — APPROVED** [review]
LGTM -- will merge once the linux CI builds succeed.

**Pull request merged**